File: experiments/benchmarks/microbench_launch.c
/*
 * microbench_launch.c — raw OpenCL per-kernel-launch overhead baseline.
 *
 * 10000회 noop 커널 launch 시간을 측정하여 per-launch 오버헤드(μs)를 구한다.
 * Rust ocl::core 버전과 동일한 구조로 작성해 wrapper 오버헤드를 비교한다.
 *
 * Android 빌드:
 *   $TOOLCHAIN/bin/aarch64-linux-android28-clang -O3 \
 *     -o microbench_launch_c microbench_launch.c -lOpenCL
 *
 * 실행 (LD_LIBRARY_PATH로 /vendor/lib64/libOpenCL.so 또는 /data/local/tmp 지정):
 *   LD_LIBRARY_PATH=/data/local/tmp ./microbench_launch_c
 */

#define CL_TARGET_OPENCL_VERSION 200
#include <CL/cl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define N 10000
#define WARMUP 1000

static const char *kernel_src =
    "__kernel void noop(__global int *dummy) {\n"
    "    (void)dummy;\n"
    "}\n";

static void die(const char *msg, cl_int err) {
    fprintf(stderr, "FATAL: %s (cl_int=%d)\n", msg, (int)err);
    exit(1);
}

static double now_us(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec * 1e6 + (double)ts.tv_nsec / 1e3;
}

int main(void) {
    cl_int err;

    /* 1. Platform + Device */
    cl_platform_id platform;
    err = clGetPlatformIDs(1, &platform, NULL);
    if (err != CL_SUCCESS) die("clGetPlatformIDs", err);

    cl_device_id device;
    err = clGetDeviceIDs(platform, CL_DEVICE_TYPE_GPU, 1, &device, NULL);
    if (err != CL_SUCCESS) die("clGetDeviceIDs", err);

    char pname[256] = {0}, dname[256] = {0};
    clGetPlatformInfo(platform, CL_PLATFORM_NAME, sizeof(pname), pname, NULL);
    clGetDeviceInfo(device, CL_DEVICE_NAME, sizeof(dname), dname, NULL);
    printf("C raw OpenCL microbench\n");
    printf("Platform: %s\n", pname);
    printf("Device:   %s\n", dname);

    /* 2. Context + Queue (profiling OFF) */
    cl_context ctx = clCreateContext(NULL, 1, &device, NULL, NULL, &err);
    if (err != CL_SUCCESS) die("clCreateContext", err);

    /* production 경로와 동일: profiling flag OFF */
    cl_command_queue queue =
#if defined(CL_VERSION_2_0)
        clCreateCommandQueueWithProperties(ctx, device, NULL, &err);
#else
        clCreateCommandQueue(ctx, device, 0, &err);
#endif
    if (err != CL_SUCCESS) die("clCreateCommandQueue", err);

    /* 3. Program + Kernel */
    const char *srcs[1] = {kernel_src};
    size_t lens[1] = {strlen(kernel_src)};
    cl_program prog = clCreateProgramWithSource(ctx, 1, srcs, lens, &err);
    if (err != CL_SUCCESS) die("clCreateProgramWithSource", err);

    err = clBuildProgram(prog, 1, &device, NULL, NULL, NULL);
    if (err != CL_SUCCESS) {
        size_t log_size = 0;
        clGetProgramBuildInfo(prog, device, CL_PROGRAM_BUILD_LOG, 0, NULL, &log_size);
        char *log = (char *)malloc(log_size + 1);
        clGetProgramBuildInfo(prog, device, CL_PROGRAM_BUILD_LOG, log_size, log, NULL);
        log[log_size] = 0;
        fprintf(stderr, "Build log:\n%s\n", log);
        free(log);
        die("clBuildProgram", err);
    }

    cl_kernel kernel = clCreateKernel(prog, "noop", &err);
    if (err != CL_SUCCESS) die("clCreateKernel", err);

    /* 4. Dummy buffer */
    cl_mem dummy_buf =
        clCreateBuffer(ctx, CL_MEM_READ_WRITE, sizeof(cl_int), NULL, &err);
    if (err != CL_SUCCESS) die("clCreateBuffer", err);

    /* 5. Warmup */
    size_t gws[1] = {1};
    for (int i = 0; i < WARMUP; ++i) {
        err = clSetKernelArg(kernel, 0, sizeof(cl_mem), &dummy_buf);
        if (err != CL_SUCCESS) die("clSetKernelArg(warmup)", err);
        err = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, gws, NULL, 0, NULL, NULL);
        if (err != CL_SUCCESS) die("clEnqueueNDRangeKernel(warmup)", err);
    }
    clFinish(queue);

    /* 6. Measured region */
    double t_start = now_us();
    for (int i = 0; i < N; ++i) {
        err = clSetKernelArg(kernel, 0, sizeof(cl_mem), &dummy_buf);
        if (err != CL_SUCCESS) die("clSetKernelArg(measure)", err);
        err = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, gws, NULL, 0, NULL, NULL);
        if (err != CL_SUCCESS) die("clEnqueueNDRangeKernel(measure)", err);
    }
    clFinish(queue);
    double t_end = now_us();

    double total_us = t_end - t_start;
    double per_us = total_us / (double)N;

    printf("N = %d\n", N);
    printf("total = %.2f ms\n", total_us / 1000.0);
    printf("per-launch = %.3f us (includes clSetKernelArg + clEnqueueNDRangeKernel + avg clFinish/N)\n",
           per_us);
    printf("(kernel GPU time per call estimate: negligible, noop)\n");

    /* Cleanup */
    clReleaseMemObject(dummy_buf);
    clReleaseKernel(kernel);
    clReleaseProgram(prog);
    clReleaseCommandQueue(queue);
    clReleaseContext(ctx);
    return 0;
}

File: engine/src/bin/microbench_launch.rs
//! microbench_launch — OpenCL per-kernel-launch overhead 측정
//!
//! 목적: Rust ocl::core API를 통해 noop 커널을 10000회 launch 하여
//! per-launch 오버헤드(μs)를 실측한다. 이 값이 Decode 갭(30.6 ms/tok,
//! 364 launches/token)을 설명하는지 검증한다.
//!
//! 측정 원칙:
//! - 기존 `OpenCLBackend` 사용 금지 (오버헤드 배제)
//! - `CL_QUEUE_PROFILING_ENABLE` OFF (production dispatch 경로와 동일)
//! - set_kernel_arg + enqueue_kernel만 루프 (profile_events 없음)
//! - wall-clock `Instant`로 전체 구간 측정

#[cfg(not(feature = "opencl"))]
fn main() {
    eprintln!("microbench_launch requires --features opencl");
    std::process::exit(2);
}

#[cfg(feature = "opencl")]
fn main() -> anyhow::Result<()> {
    use ocl::core::{ArgVal, Event, Mem};
    use ocl::{Context, Device, Platform, Program, Queue};
    use std::time::Instant;

    // === 1. Platform / Device / Context / Queue ===
    let platform = Platform::default();
    let device = Device::first(platform)?;
    let platform_name = platform.name().unwrap_or_else(|_| "unknown".into());
    let device_name = device.name().unwrap_or_else(|_| "unknown".into());
    println!("Rust ocl::core microbench");
    println!("Platform: {}", platform_name);
    println!("Device:   {}", device_name);

    let context = Context::builder()
        .platform(platform)
        .devices(device)
        .build()?;

    // production 경로와 동일: profiling flag OFF
    let queue = Queue::new(&context, device, None)?;

    // === 2. Compile noop kernel ===
    let src = r#"
        __kernel void noop(__global int *dummy) {
            // arg 바인딩만 검증; 실제로 쓰지 않음
            (void)dummy;
        }
    "#;
    let program = Program::builder()
        .devices(device)
        .src(src)
        .build(&context)?;

    let kernel = ocl::core::create_kernel(&program, "noop")?;

    // === 3. Dummy buffer (cl_int[1]) ===
    let dummy_buf: Mem = unsafe {
        ocl::core::create_buffer::<_, i32>(context.as_core(), ocl::core::MEM_READ_WRITE, 1, None)?
    };

    // === 4. Warmup: 1000 launches + finish ===
    let gws = [1usize, 1, 1];
    let lws: Option<[usize; 3]> = None; // driver picks local size
    for _ in 0..1000 {
        unsafe {
            ocl::core::set_kernel_arg(&kernel, 0, ArgVal::mem(&dummy_buf))?;
            ocl::core::enqueue_kernel(
                &queue,
                &kernel,
                1,
                None,
                &gws,
                lws,
                None::<&Event>,
                None::<&mut Event>,
            )?;
        }
    }
    ocl::core::finish(&queue)?;

    // === 5. Measured region: 10000 launches ===
    const N: usize = 10_000;
    let t_start = Instant::now();
    for _ in 0..N {
        unsafe {
            ocl::core::set_kernel_arg(&kernel, 0, ArgVal::mem(&dummy_buf))?;
            ocl::core::enqueue_kernel(
                &queue,
                &kernel,
                1,
                None,
                &gws,
                lws,
                None::<&Event>,
                None::<&mut Event>,
            )?;
        }
    }
    ocl::core::finish(&queue)?;
    let elapsed = t_start.elapsed();

    let total_us = elapsed.as_secs_f64() * 1_000_000.0;
    let per_launch_us = total_us / N as f64;

    println!("N = {}", N);
    println!("total = {:.2} ms", total_us / 1000.0);
    println!(
        "per-launch = {:.3} us (includes set_kernel_arg + enqueue_kernel + avg finish/N)",
        per_launch_us
    );
    println!("(kernel GPU time per call estimate: negligible, noop)");

    Ok(())
}
